src/vsr/superblock_free_set.zig

```diff
@@ -307,15 +367,45 @@ pub const FreeSet = struct {
     ///        (Note: This must be careful not to release while any reservations are held
     ///        to avoid making the reservation's acquire()s nondeterministic).
     pub fn release(set: *FreeSet, address: u64) void {
+        assert(set.opened);
         const block = address - 1;
         assert(set.blocks.isSet(block));
         assert(!set.staging.isSet(block));
 
         set.staging.set(block);
     }
 
+    /// Mark the given addresses as allocated in the current checkpoint, but free in the next one.
+    ///
+    /// This is used only when reading a free set from the grid. On disk representation of the
+    /// free set doesn't include the blocks storing the free set itself, and these blocks must be
+    /// manually patched in after decoding. As the next checkpoint will have a completely different
+    /// free set, the blocks can be simultaneously released.
+    fn mark_released(set: *FreeSet, addresses: []const u64) void {
+        assert(!set.opened);
+        var address_previous: u64 = 0;
+        for (addresses) |address| {
+            assert(address > 0);
+
+            // Assert that addresses are sorted and unique. Sortedness is not a requirement, but
+            // a consequence of "first free" allocation algorithm.
+            assert(address > address_previous);
+            address_previous = address;
+
+            const block = address - 1;
+            assert(!set.blocks.isSet(block));
+            assert(!set.staging.isSet(block));
+
+            const shard = @divFloor(block, shard_bits);
+            if (set.find_free_block_in_shard(shard) == null) set.index.set(shard);
```

Comment:
Oh dear, this is actually a bug! I should _first_ set the bit in the `blocks`, and only afterwards, if that fills up the shard, mark the index.

Reply:
Hm, curiously, the following diff

```
diff --git a/src/vsr/superblock_free_set.zig b/src/vsr/superblock_free_set.zig
index 3ef3db6c..b9250b54 100644
--- a/src/vsr/superblock_free_set.zig
+++ b/src/vsr/superblock_free_set.zig
@@ -325,9 +325,9 @@ pub const FreeSet = struct {
         assert(!set.blocks.isSet(block));
         assert(!set.staging.isSet(block));
 
-        set.blocks.set(block);
         // Update the index when every block in the shard is allocated.
         if (set.find_free_block_in_shard(shard) == null) set.index.set(shard);
+        set.blocks.set(block);
 
         const address = block + 1;
         return address;
@@ -399,9 +399,9 @@ pub const FreeSet = struct {
             assert(!set.staging.isSet(block));
 
             const shard = @divFloor(block, shard_bits);
-            set.blocks.set(block);
             // Update the index when every block in the shard is allocated.
             if (set.find_free_block_in_shard(shard) == null) set.index.set(shard);
+            set.blocks.set(block);
             set.staging.set(block);
         }
     }
@@ -426,6 +426,11 @@ pub const FreeSet = struct {
         assert(set.reservation_count == 0);
         assert(set.reservation_blocks == 0);
 
+        var shard: usize = 0;
+        while (shard < set.index.bit_length) : (shard += 1) {
+            assert(set.index.isSet(shard) == (set.find_free_block_in_shard(shard) == null));
+        }
+
         var it = set.staging.iterator(.{ .kind = .set });
         while (it.next()) |block| {
             set.staging.unset(block);
```

trips neither vopr nor forest fuzz...

Reply:
:thinking: I would understand the VOPR not hitting it, but the forest fuzzer is surprising. Shards are 4096 blocks.

Reply:
In `FreeSet.acquire`:

```zig
        const shard = find_bit(
            set.index,
            @divFloor(reservation.block_base, shard_bits),
            div_ceil(reservation.block_base + reservation.block_count, shard_bits),
            .unset,
        ) orelse return null;
        assert(!set.index.isSet(shard));

        const reservation_start = @max(
            shard * shard_bits,
            reservation.block_base,
        );
        const reservation_end = reservation.block_base + reservation.block_count;
        const block = find_bit(
            set.blocks,
            reservation_start,
            reservation_end,
            .unset,
        ) orelse return null;
```

I think we aren't maintaining the shard index strictly! i.e. if we picked a `shard` which is (unexpectedly) full, that would silently work, as long as it wasn't the last shard covered by the reservation. Which it usually wouldn't be, since reservations are worst-case and typically much larger than necessary.

Reply:
Yeah, it makes sense that just not updating the index precisely won’t blow up immediately (it could still explode when a replica restarts and gets a different, precise index by reading from disk). That’s why in the diff above I added a loop which asserts, at a checkpoint, that the index is consistent with blocks. That one I would expect to get tripped…

will look more into this tomorrow…

Reply:
Mystery solved, I think! 

The reason why this isn't tripped by normal fuzzers is that the shard size, 4096, is pretty large, and we need to fill all of shard's bits to hit this. 

It _is_ tiggered by free_set_fuzz though, via https://github.com/tigerbeetle/tigerbeetle/pull/1319

Thanks again for noticing that something's off here dj, this would have been a horrible bug to discover after the fact.